subsystems/FPIDSubsystem.java

```diff
@@ -0,0 +1,140 @@
+import org.frc5587.lib.pid.PID;
+import edu.wpi.first.wpilibj.SpeedControllerGroup;
+import edu.wpi.first.wpilibj.controller.PIDController;
+import edu.wpi.first.wpilibj.SpeedController;
+import edu.wpi.first.wpilibj2.command.PIDSubsystem;
+
+public abstract class FPIDSubsystem extends PIDSubsystem {
+    protected FPIDConstants constants;
+    protected SpeedController[] motors;
+    protected SpeedControllerGroup motorGroup;
+
+    // what type of value we can get from an encoder
+    public static enum EncoderValueType {
+        Velocity, Position
+    }
+
+    public static class FPIDConstants {
+        public double speedMultiplier, gearing;
+        public int encoderCPR;
+        public PID pid;
+        public FFPIDController ff;
+
+        public FPIDConstants(double speedMultiplier, double gearing, int encoderCPR, PID pid, FFPIDController ff) {
+            this.speedMultiplier = speedMultiplier;
+            this.gearing = gearing;
+            this.encoderCPR = encoderCPR;
+            this.pid = pid;
+            this.ff = ff;
+        }
+    }
+
+    /** 
+    * pass motors as a SpeedControllerGroup 
+    */
+    public FPIDSubsystem(FPIDConstants constants, SpeedControllerGroup motorGroup) {
+        super(new PIDController(constants.pid.kP, constants.pid.kI, constants.pid.kD));
+        /**
+        * enable PID control when starting
+        */
+        this.enable();
+
+        this.constants = constants;
+        this.motorGroup = motorGroup;
+
+        configureMotors();
+    }
+
+    /**
+    * have the implementing class configure the motors 
+    */
+    public abstract void configureMotors();
+
+    /**
+    * the implementing class also needs to get and set encoder values
+    * @param type what type of value we're getting from the encoder. values Position and Velocity as defined above in EncoderValueType
+    */
+    public abstract double getEncoderValue(EncoderValueType type);
+    public abstract void setEncoderPosition(double position);
+
+    /**
+    * the implementing class should decide how to calculate Feedforward 
+    * because we don't know what model to use
+    */
+    public abstract double calcFeedForward(double position, double velocity);
+    public abstract double calcFeedForward();
+
+    /**
+    * it should also do getMeasurement because we don't know what measurement to use
+    * (measurement could be an angle, height, etc)
+    */
+    @Override
+    public abstract double getMeasurement();
+
+    @Override
+    public abstract void periodic();
+
+    /**
+    * move the mechanism based on a given throttle 
+    */
+    public void moveByThrottle(double throttle) {
+        motorGroup.set(throttle * constants.speedMultiplier);
+    }
+
+    /**
+    *  move the mechanism based on a constant multiplier (for operation with buttons)
+    */
+     public void moveByFixedSpeed() {
+        motorGroup.set(constants.speedMultiplier);
+    }
+
+    public void moveFixedReversed() {
```

Comment:
remove these, this is stuff that would be added for specific classes, and tbh, for an ffpid subsystem, it likely will never be useful